onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc

```diff
@@ -551,13 +555,31 @@ Status ProcessLogits(const OrtValue& logits,                                 //
   gsl::span<const int32_t> next_tokens(cpu_state->topk_tokens.data(), beam_state->next_tokens.size());
   gsl::span<const int32_t> next_indices(cpu_state->topk_indices.data(), beam_state->next_indices.size());
 
-  // Limitation: beam scorer runs in CPU. It might be better to use CUDA kernel to replace it.
+  // TODO: Implement BeamScorer on CUDA
   beam_scorer->Process(
       sequences,
       next_scores,
       next_tokens,
       next_indices);
 
+  // TODO: This is a temporary work-around as BeamScorer currently only runs on CPU.
+  // We can remove these kinds of work-arounds once BeamScorer runs on CUDA eventually.
+  auto chosen_indices = beam_scorer->GetNextIndices();
+  auto beam_state_chosen_indices = beam_state->chosen_indices;
+
+  if (!beam_state_chosen_indices.empty()) {
```

Comment:
This will be non-empty ONLY if we use DecoderMaskedMultiheadAttention and this copy will not kick in for the regular Attention based BeamSearch